src/keystore.ts

```diff
@@ -0,0 +1,131 @@
+import uuid from "uuid";
+import { scrypt, PBKDF2, DefaultPBKDF2Params, AES_128_CTR, SHA256, DefaultScryptParams } from "./utils/crypto";
+import { PublicKey, PrivateKey } from "@chainsafe/bls";
+import * as random from "secure-random";
+import { IKeystoreCrypto, ScryptParams, PBKDF2Params, bytes, IKeystore, IKeystoreParams } from ".";
+import { KeystoreCrypto } from "./keystore-crypto";
+import { Buffer } from "buffer";
+
+export class Keystore implements IKeystore {
+  public crypto: IKeystoreCrypto = new KeystoreCrypto({});
+  public readonly pubkey: string = "";
+  public readonly path: string = "";
+  public readonly uuid: string = uuid.v4();
+  public readonly version: number = 4;
+
+  private kdf(password: string, args: ScryptParams | PBKDF2Params): Buffer {
+    switch(this.crypto.kdf.function){
+      case "scrypt":
+        return scrypt(password, args as ScryptParams);
+      case "pbkdf2":
+        return PBKDF2(password, args as PBKDF2Params);
+      default:
+        throw new Error("Unsupported crypto function");
+    }
+  }
+
+  constructor(keystore?: IKeystoreParams){
+    if(keystore){
+      this.crypto = new KeystoreCrypto(keystore.crypto);
+      this.pubkey = keystore.pubkey || "";
+      this.path = keystore.path || "";
+      this.uuid = keystore.uuid || uuid.v4();
+      this.version = keystore.version || 4;
+    }
+  }
+
+  public static fromJson(json: Record<string, any>): Keystore {
```

Comment:
again, you will receive string here and you have to call JSON.parse on that string and cast resulted object IKeystoreParams and call constructor.

On other submodules it's not required but if you have it it has to do the same thingas here.